Mathlib/LinearAlgebra/PiTensorProduct/Generators.lean

```diff
@@ -0,0 +1,150 @@
+  apply (equivTensorPiTensorComplSingleton R M i₀).injective
+  simp only [LinearEquiv.apply_symm_apply, equivTensorPiTensorComplSingleton_tprod,
+    Function.complSingletonLift_self]
+  congr
+  ext ⟨i, hi⟩
+  rw [Function.complSingletonLift_of_neq]
+
+end equivTensorPiTensorComplSingleto
+
+variable {R} {ι : Type*} [Finite ι]
+  {M : ι → Type*} [∀ i, AddCommGroup (M i)] [∀ i, Module R (M i)]
+  {N : Type*} [AddCommGroup N] [Module R N]
+  {γ : ι → Type*} {g : ⦃i : ι⦄ → (j : γ i) → M i}
+
+lemma ext_of_span_eq_top
+    (hg : ∀ i, Submodule.span R (Set.range (@g i)) = ⊤)
+    {φ φ' : (⨂[R] i, M i) →ₗ[R] N}
+    (h : ∀ (j : (i : ι) → γ i),
+      φ (tprod _ (fun i ↦ g (j i))) = φ' (tprod _ (fun i ↦ g (j i)))) :
+    φ = φ' := by
+  obtain ⟨n, hn⟩ : ∃ (n : ℕ), Nat.card ι = n := ⟨_, rfl⟩
+  revert ι
+  induction n with
+  | zero =>
+    intro ι _ M _ _ γ g _ φ φ' h hι
+    ext x
+    have : IsEmpty ι := by
+      rw [Nat.card_eq_zero] at hι
+      obtain (_ | h) := hι
+      · assumption
+      · exfalso
+        rw [← not_finite_iff_infinite] at h
+        exact h inferInstance
+    obtain rfl : x = fun i ↦ @g i (by apply isEmptyElim) := by
+      ext i
+      apply @isEmptyElim (a := i) _ _
+    apply h
+  | succ n hn =>
+    intro ι _ M _ _ γ g hg φ φ' h hι
+    have : Nonempty ι := ((Nat.card_pos_iff (α := ι)).1 (by omega)).1
+    have i₀ : ι := Classical.arbitrary _
+    let e := equivTensorPiTensorComplSingleton R M i₀
+    obtain ⟨ψ, rfl⟩ : ∃ ψ, φ = LinearMap.comp ψ e.toLinearMap :=
+      ⟨φ.comp e.symm.toLinearMap, by ext; simp⟩
+    obtain ⟨ψ', rfl⟩ : ∃ ψ', φ' = LinearMap.comp ψ' e.toLinearMap :=
+      ⟨φ'.comp e.symm.toLinearMap, by ext; simp⟩
+    dsimp [e] at h
+    congr 1
+    apply (TensorProduct.lift.equiv _ _ _ _).symm.injective
+    rw [Submodule.linearMap_eq_iff_of_span_eq_top _ _ (hg i₀)]
+    rintro ⟨_, ⟨g₀, rfl⟩⟩
+    apply hn (g := fun i (j : γ i.1) ↦ by exact g j)
+    · intro i
+      exact hg _
+    · intro j
+      classical
+      simp only [lift.equiv_symm_apply]
+      convert h (Function.complSingletonLift i₀ j g₀) using 1
+      all_goals
+        simp only [equivTensorPiTensorComplSingleton_tprod,
+          Function.complSingletonLift_self]
+        congr
+        ext x
+        congr
+        rw [Function.complSingletonLift_of_neq]
+    · exact Nat.card_singleton_compl i₀ hι
+
+lemma _root_.MultilinearMap.ext_of_span_eq_top
+    (hg : ∀ i, Submodule.span R (Set.range (@g i)) = ⊤)
+    {φ φ' : MultilinearMap R M N}
+    (h : ∀ (j : (i : ι) → γ i), φ (fun i ↦ g (j i)) = φ' (fun i ↦ g (j i))) :
+    φ = φ' := by
+  suffices lift φ = lift φ' by
+    ext m
+    simpa using DFunLike.congr_fun this (tprod _ m)
+  exact PiTensorProduct.ext_of_span_eq_top hg (fun j ↦ by simpa using h j)
+
+lemma submodule_span_eq_top
+    (hg : ∀ i, Submodule.span R (Set.range (@g i)) = ⊤) :
+    Submodule.span R (Set.range (fun j : ((i : ι) → γ i) ↦
+      ⨂ₜ[R] (i : ι), g (j i))) = ⊤ := by
```

Comment:
Is there a more general statement here that expresses `Submodule.span R (Set.range (fun j : ((i : ι) → γ i) ↦⨂ₜ[R] (i : ι), g (j i)))` in terms of `Submodule.span R (Set.range (@g i))`?

Reply:
I do not see any such obvious statement with the current definitions. In order to do this, we would have to define a "PiTensorProduct" operation on submodules, but I am not sure we need/want to do this; and this would have to be in separate PRs anyway.